### DISCUSSION TITLE
ROS2 Control For VESC (motor controller) not working with two VESCs

### DISCUSSION THREAD

**Discussion by Persius48** [prl-mushr/mushr#115] [Q&A] [answered]
Hello Evryone
I am currently working on a project to make a four-wheeled differential drive robot using 4 VESCs with ROS2 as the means of communication. Two vesc will be masters and the rest two will be slaves. I found the [vesc_hw_interface](https://github.com/sbgisen/vesc/tree/humble-devel/vesc_hw_interface) in which I used for one VESC and it was okay. But when I tried to use two VESCs with some modification in the ros2_control_xacro by including two hardware and two joints, there was an error like the following

> Invalid checksum [ros2_control_node-1] [ERROR] [1693755709.123338959] [VescHwInterface]: Out-of-sync with VESC, discarding 49 bytes. [ros2_control_node-1] [ERROR] [1693755709.128419540] [VescHwInterface]: Invalid end-of-frame character [ros2_control_node-1] [ERROR] [1693755709.128438065] [VescHwInterface]: Invalid payload length [ros2_control_node-1] [ERROR] [1693755709.128448735] [VescHwInterface]: Out-of-sync with VESC, discarding 21 bytes. [ros2_control_node-1] [ERROR] [1693755709.163161253] [VescHwInterface]: Invalid end-of-frame character [ros2_control_node-1] [ERROR] [1693755709.163192823] [VescHwInterface]: Out-of-sync with VESC, discarding 8 bytes. [ros2_control_node-1] [ERROR] [1693755709.203127802] [VescHwInterface]: Invalid end-of-frame character [ros2_control_node-1] [ERROR] [1693755709.203179020] [VescHwInterface]: Out-of-sync with VESC, unknown data leading valid frame. Discarding 1 bytes.

I tried using the controllers (joint_state_broadcaster and velocity_controller), which worked with a single VESC, launching it with two VESCs with two ros2_control tags including two separate actuator-type hardware. I also tried to change the velocity controller to diff_drive_controller and tried the same thing with two VESCs. But it fails every time. I am attaching the screenshot of my configuration and xacro file. Any suggestion will be really helpful.
![1](https://github.com/prl-mushr/mushr/assets/139690589/35c25a63-9bb5-43c6-a8e2-6f0a46c70d6d)
![2](https://github.com/prl-mushr/mushr/assets/139690589/44083be5-be97-4c7f-a314-8673988ed887)
![3](https://github.com/prl-mushr/mushr/assets/139690589/f5ab763e-ae95-43d4-8825-e11d2f73857f)
![4](https://github.com/prl-mushr/mushr/assets/139690589/fe54ce9a-3bbf-4475-a12c-54e6d68b09ee)
![5](https://github.com/prl-mushr/mushr/assets/139690589/b3f13989-27cd-4fe5-9f80-4fcda58aa830)
![6](https://github.com/prl-mushr/mushr/assets/139690589/6378062f-44b5-48e7-a6b0-89ff4ded8c7c)

[NB. I commented out the servo parameters and used parameters that are required for velocity as command interface and positional and velocity for state interfaces]

**Comment by sidtalia** [comment, marked as answer]
Hi, Apologies for the late response, we do not currently support ROS2. However, as a hunch, the two vescs have different port names. If you execute 
```bash
ls /dev/ttyACM*
```
you should probably see 
```bash
/dev/ttyACM0
/dev/ttyACM1
```
You would then need to specify the port name for each vesc (usually given in the launch file). 

Thanks,
Regards,
Sidharth.
